icontrol/authtoken.py

```diff
@@ -0,0 +1,137 @@
+                                 json=login_body,
+                                 verify=False,
+                                 auth=HTTPBasicAuth(self.username,
+                                                    self.password))
+        self.attempts += 1
+        if not response.ok or not hasattr(response, "json"):
+            error_message = '%s Unexpected Error: %s for uri: %s\nText: %r' %\
+                            (response.status_code,
+                             response.reason,
+                             response.url,
+                             response.text)
+            raise iControlUnexpectedHTTPError(error_message,
+                                              response=response)
+        respJson = response.json()
+
+        expiration_bigip, created_bigip = None, None
+        try:
+            self.token = respJson['token']['token']
+            expiration_bigip = int(respJson['token']['expirationMicros']) / \
+                1000000.0
+            created_bigip = int(respJson['token']['lastUpdateMicros']) / \
+                1000000.0
+        except KeyError:
+            error_message = \
+                '%s Unparseable Response: %s for uri: %s\nText: %r' %\
+                (response.status_code,
+                 response.reason,
+                 response.url,
+                 response.text)
+            raise iControlUnexpectedHTTPError(error_message,
+                                              response=response)
+
+        # Set our token expiration time.
+        # The expirationMicros field is when BIG-IP will expire the token
+        # relative to its local clock.  To avoid issues caused by incorrect
+        # clocks or network latency, we'll compute an expiration time that is
+        # referenced to our local clock, and expires slightly before the token
+        # should actually expire on BIG-IP
+
+        # Reference to our clock: compute for how long this token is valid as
+        # the difference between when it expires and when it was created,
+        # according to BIG-IP.
+        if expiration_bigip < created_bigip:
+            error_message = \
+                '%s Token already expired: %s for uri: %s\nText: %r' % \
+                (response.status_code,
+                 time.ctime(expiration_bigip),
+                 response.url,
+                 response.text)
+            raise iControlUnexpectedHTTPError(error_message,
+                                              response=response)
+        valid_duration = expiration_bigip - created_bigip
+
+        # Assign new expiration time that is 1 minute earlier than BIG-IP's
+        # expiration time, as long as that would still be at least a minute in
+        # the future.  This should account for clock skew between us and
+        # BIG-IP.  By default tokens last for 60 minutes so getting one every
+        # 59 minutes instead of 60 is harmless.
+        if valid_duration > 120.0:
```

Comment:
If a failed request is the worst outcome, then I don't think we need additional heuristics.

In the case where the BigIP considers a token expired, and a client does not...   could an attacker (that already knows the victim's password) request a new token, get it, and effectively DOS the victim?   I guess not because it seems that the BigIP will issue a new token even if the requester already holds an unexpired token?

OK, I've talked myself into seeing things your way.